### ISSUE TITLE
Getting "string index out of range" error while trying to train

### ISSUE THREAD

**Issue by AshokR** [datquocnguyen/RDRPOSTagger#1]
Thanks a lot for making available this open source package!

I am trying to train RDRPOSTagger for POS tagging of Tamil language. I have converted the POS tagged text to be same format as your goldTrain file. Each line in my input training corpus is a tokenized/word-segmented sentence. I was able to train using a small test file (after  some trial and error). I was able to generate the .RDR and .DICT files.

However, when running a larger file, I am getting a "string index out of range" error. This happens immediately after starting to generate the lexicon. So, the next 3 steps, namely extracting the raw text corpus, POS tagging and learning the tree model of rules do not happen. I am using a Ubuntu Linux box.The command I am using to train is: python RDRPOSTagger.py train [my POS tagged formatted file path and name]

Any tips on how to identify the source of this error will be very helpful.

**Comment by datquocnguyen** [comment]
``Each line in my input training corpus is a tokenized/word-segmented sentence.``

I guess you mean each line be a sequence of ``word/tag`` pairs.

Anyway, I am pretty sure that there is something incorrectly formatted in the larger training file. 

For example, in some sentence, you have a word without a POS tag attached. (That would be possible when some annotator forgot to put a POS label for some word.) So you got -1 index for finding character "/", and then a "string index out of range" error. 

What you should do is just to uncomment the code line 36: ``# print i`` in ``Utility/LexiconCreator.py``. Then you can know which the incorrectly formatted sentence (sentence id) in training data is. And then you are able to fix the data.

**Comment by AshokR** [comment]
Thanks for your very quick response :)
I am stumped. Even this simple three line Tamil text file is causing the "string index out of range" error. This happens immediately after starting to generate the lexicon. I uncommented line 36 in Utility/LexiconCreator.py as suggested by you. It printed out 0, 1 and 2. However, I cannot see any missing tags or anything else that could be causing the problem.
நல்ல/ADJ தலைவர்/NN என்பவர்/COM யாராக/ADV இருக்க/VINT முடியும்/VAX ./. 
ஒரு/DET அப்துல்கலாம்/NNP மட்டும்/PPO போதாது/VF ./. 
அவரும்/PRP இன்று/ADV நம்மிடம்/PRP இல்லை/VAX ./.

**Comment by datquocnguyen** [comment]
I did test with three test sentences you mentioned. And I found an error of "list index out of range". (It is not "string index out of range". And I am sure that such error "string index out of range" comes from some incorrectly formatted sentence in training data.)

Regarding to the error of "list index out of range", I identified the error source. It is because ``LexiconCreator.py`` could not find an upper case character in the training set. So I did revise some code lines so that ``LexiconCreator.py`` can handle this issue. Now the program enables to run on the three test sentences.

I just updated the source codes, and you can download it again from github.

If there is any problem, please let me know. Thanks a lot :)

**Comment by AshokR** [comment]
You are right. The error for the above sample data was "list index out of range". Thanks for fixing that. Tamil language doesn't have any upper case letters! 

As you pointed out, my training corpus has several tagging errors that are causing "string index out of range" error. I am working on fixing them.

However, uncommenting the code line 36: # print i in Utility/LexiconCreator.py is not helpful. It is just printing all the line numbers!

**Comment by datquocnguyen** [comment]
@AshokR I just updated two code lines in `LexiconCreator.py`, now it can print out all the annotation errors in your training corpus. Redownload from github to obtain the updated version.
